metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala

```diff
@@ -1,326 +1,78 @@
-import scala.util.Random
-import scala.util.{Failure, Success}
-import scala.util.control.NonFatal
-import scala.util.Try
+import bloop.launcher.LauncherMain
+import java.nio.charset.StandardCharsets
+import bloop.bloopgun.core.Shell
+import java.nio.channels.Channels
+import java.nio.channels.Pipe
 
 /**
- * Establishes a connection with a bloop server.
+ * Establishes a connection with a bloop server using Bloop Launcher.
  *
  * Connects to a running bloop server instance if it is installed on the user
- * machine and has compatible version (+v1.1.0). Ignores the installed bloop
- * server instance if it is v1.0.0.
+ * machine and starts a new one if it isn't. Alternatively user can use the
+ * coursier command to launch it:
  *
- * Otherwise, if the user doesn't have bloop installed on the machine, uses
- * Coursier to fetch the jars for ch.epfl.scala:bloop-frontend and start a new
- * bloop server using classloading. A bloop server that starts via classloading
- * stops when when the metals server stops. Only metals can access the classloaded server,
- * the user cannot call it from a command-line interface.
+ * coursier launch ch.epfl.scala:bloopgun-core_2.12:{bloop-version} -- about
  *
  * Eventually, this class may be superseded by "BSP connection protocol":
  * https://github.com/scalacenter/bsp/blob/master/docs/bsp.md#bsp-connection-protocol
  */
 final class BloopServers(
     sh: ScheduledExecutorService,
     workspace: AbsolutePath,
-    client: MetalsBuildClient,
-    config: MetalsServerConfig,
-    icons: Icons,
-    embedded: Embedded,
-    statusBar: StatusBar
+    client: MetalsBuildClient
 )(implicit ec: ExecutionContextExecutorService) {
 
-  def newServer(
-      maxRetries: Int = defaultRetries
-  ): Future[Option[BuildServerConnection]] = {
-    newServerUnsafe().map(Option(_)).recoverWith {
-      case NonFatal(e) if maxRetries > 0 =>
-        val retry = defaultRetries - maxRetries + 1
-        val cause = e.getMessage
-        scribe.warn(s"BSP retry $retry due to: $cause", e)
-        newServer(maxRetries - 1)
-    }
-  }
+  def newServer(): Future[Option[BuildServerConnection]] = {
+    val launcherInOutPipe = Pipe.open()
+    val launcherIn = Channels.newInputStream(launcherInOutPipe.source())
+    val clientOut = Channels.newOutputStream(launcherInOutPipe.sink())
+
+    val clientInOutPipe = Pipe.open()
+    val clientIn = Channels.newInputStream(clientInOutPipe.source())
+    val launcherOut = Channels.newOutputStream(clientInOutPipe.sink())
+
+    val serverStarted = Promise[Unit]()
+    val launcher =
+      new LauncherMain(
+        launcherIn,
+        launcherOut,
+        System.err,
+        StandardCharsets.UTF_8,
+        Shell.default,
+        userNailgunHost = None,
+        userNailgunPort = None,
+        serverStarted
+      )
 
-  private def defaultRetries: Int =
-    if (config.bloopProtocol.isTcp) {
-      10
-    } else {
-      2
-    }
+    val job = ec.submit(new Runnable {
```

Comment:
super tiny, but this can probably be a SAM

```scala
ec.submit(_ => launcher.runLauncher(...))
```

Reply:
It would need to be something like:
```scala
    val runServer: Runnable = () =>
      launcher.runLauncher(
        BuildInfo.bloopVersion,
        skipBspConnection = false,
        Nil
      )
    val job = ec.submit(runServer)
```

I can change it, but doesn't seem like much of an improvement, because we need the explicit type there :/

Reply:
Ah, I recall @olafurpg had a similar issue, does it work if you do

```scala
val job = ec.submit[Unit](_ => launcher.runLauncher(
  BuildInfo.bloopVersion,
  skipBspConnection = false,
  Nil
))
```
?

The `Unit` annotation should help disambiguate `Runnable` from `Callable`.

Anyway, it's a super tiny thing, sorry for digressing ^^

Reply:
Doesn't work either, but thanks for the suggestion. I would love to have it simpler than what it currently is. It feels a bit too Java currently :smile:

Reply:
ahah, ok, I surrender.